Updating Carbon's safety strategy (#5914)
carbon-language/carbon-lang: docs/design/safety/terminology.md

    -   [_Safety bugs_](#safety-bugs)
    -   [_Initial bug_](#initial-bug)
    -   [_Fail-stop_](#fail-stop)
-   [Vulnerability Terminology](#vulnerability-terminology)
    -   [_Vulnerability_ or _security vulnerability_](#vulnerability-or-security-vulnerability)
    -   [_Vulnerability defense_](#vulnerability-defense)
        -   [_Detecting_](#detecting)
        -   [_Mitigating_](#mitigating)
        -   [_Preventing_ vulnerabilities](#preventing-vulnerabilities)
        -   [_Ensuring_ correctness](#ensuring-correctness)
        -   [_Hardening_](#hardening)
-   [Memory Safety Specifics](#memory-safety-specifics)
    -   [_Memory safety_](#memory-safety)
        -   [_Temporal safety_](#temporal-safety)
        -   [_Spatial safety_](#spatial-safety)
        -   [_Type safety_](#type-safety)
        -   [_Initialization safety_](#initialization-safety)
        -   [_Data-race safety_](#data-race-safety)
    -   [_Memory safety bug_](#memory-safety-bug)
    -   [_Memory-safe platform_ or _environment_](#memory-safe-platform-or-environment)
    -   [_Memory-safe language_](#memory-safe-language)

<!-- tocstop -->

## Core Terminology

### _Hazard_

> Unsafe coding construct that may lead to a bug or vulnerability.

Reviewer: (super-nit) "or vulnerability" is redundant, because all vulnerabilities are bugs...

Author: But not all bugs are vulnerabilities? The point is that in unsafe code, the bugs are much more likely to be ones that are vulnerabilities.

Reviewer: Yeah, I guess I might be thinking too mathematically here. Feel free to ignore.

> Unsafe coding construct that may lead to a bug or vulnerability.

Reviewer 2: These definitions aren't really sentences, more like noun phrases. I'm not sure they should have periods.

Author: It felt awkward to not use a period here. Originally, these were bullet points in a list. But unfortunately, I can't link to those.

Happy to have suggestions on how exactly to format and use punctuation here. I can also add the redundant but grammatical `A _hazard_ is an ...` prefix to these.

Reviewer 3: For what it's worth: the ISO drafting rules for standards say that its "Terms and Definitions" entries should be phrases, without leading capitalization or a terminating period, that can be substituted for the term in its context.

I don't think that works so well for the more elaborate definitions you give later (eg, the second sentence in "latent bug" or the qualified definition in "safety"), but I think it would help a lot for the "Vulnerability Terminology" section, where it's currently not completely clear what the link between the terminology and the definition is.

Author: > For what it's worth: the ISO drafting rules for standards say that its "Terms and Definitions" entries should be phrases, without leading capitalization or a terminating period, that can be substituted for the term in its context.
> 
> I don't think that works so well for the more elaborate definitions you give later (eg, the second sentence in "latent bug" or the qualified definition in "safety"), but I think it would help a lot for the "Vulnerability Terminology" section, where it's currently not completely clear what the link between the terminology and the definition is.

Would that work for you with the rendered format of markdown headers?

I'd be happy to use either a substitutable phrase or full sentences if so?

Reviewer 3: I think we can iterate on this separately from this proposal.

### _Bug_ or _defect_

Reachable program behavior contrary to the author's intent.

> #### _Patent_ or _active bug_

Reviewer 2: Do we need two terms here?
```suggestion
#### _Active bug_
```

Buggy behavior that is actively occurring for users of the program.

#### _Latent bug_

> Buggy behavior that does not currently occur for users, but can occur. Behaviors

Reviewer: Is the intention for "can occur" to have the same definition as "is reachable"? 

It would be easier to understand if you'd use the same words or explicitly explain the difference.

Author: Sure. Added some words to explain this a bit more.

that can happen but don't today in practice _are always still bugs_!

### _Safety_

Absent a qualifier or narrow context, refers to _system safety_, and _safety
engineering_. Always a property of a system or product as a whole, including
human factors, etc.

### _Code_, _software_, or _program safety_

Invariants or limits on program behavior in the face of bugs.

-   Very narrow and specific meaning.
-   Often necessary but not sufficient for system safety.

### _Safety bugs_

Bugs where some aspect of program behavior has insufficient (often none)
invariants or limits.

For example, _undefined behavior_ definitionally has no invariant or limit, and

> is always a safety bug.

Reviewer 4: ```suggestion
reaching it is always a safety bug.
```

### _Initial bug_

The first behavior contrary to the author's intent, distinct from subsequent
deviations.

### _Fail-stop_

The behavior of immediately terminating the program, minimizing any further
business logic. This is in contrast to any form of "correct" program
termination, continuing execution, or unwinding.

## Vulnerability Terminology

### _Vulnerability_ or _security vulnerability_

> A subset of bugs that creates the possibility for a malicious actor to subvert a

Reviewer: (super-nit) Technically, a vulnerability is not a subset of bugs. It's a bug. The set of vulnerabilities is a subset of the set of bugs, but a single vulnerability is not a set.

Author: Can you suggest an edit that would clarify this effectively? I'm not really sure what edit would help here TBH...

Reviewer: ```suggestion
A bug that creates the possibility for a malicious actor to subvert a
```

Author: Oh, I see the confusion. Sure. =]

be exploited.

#### _Preventing_ vulnerabilities

While still a bug, making it impossible to be a vulnerability. Often this is
done by defining behavior to [fail-stop](#fail-stop).

#### _Ensuring_ correctness

No longer a bug, much less a vulnerability.

#### _Hardening_

Combinations of mitigation, prevention, and ensured correctness to reduce the
practical risk of vulnerabilities due to bugs.

## Memory Safety Specifics

### _Memory safety_

Having well-defined and predictable behavior regarding memory access, even in
the face of bugs. Memory safety encompass several key aspects:

#### _Temporal safety_

Memory accesses occur only within the valid lifetime of the intended memory
object.

#### _Spatial safety_

Memory accesses remain within the intended bounds of memory regions.

#### _Type safety_

Memory is accessed and interpreted according to its intended type, preventing
type confusion.

#### _Initialization safety_

Memory is properly initialized before being read, avoiding the use of
uninitialized data.

#### _Data-race safety_

Memory writes are synchronized with reads or writes on other threads.

### _Memory safety bug_

A safety bug that violates memory safety.

> ### _Memory-safe platform_ or _environment_

Reviewer: I think an example would be useful. Are you referring to something like the JVM?

Author: No, more a memory safe sandbox. Added an example.

> ### _Memory safe platform_ or _environment_

Reviewer 4: ```suggestion
### _Memory-safe platform_ or _environment_
```

A computing platform or execution environment that provides mechanisms to
prevent memory safety bugs in programs running on it from becoming
vulnerabilities. This is a _systems_ path to achieving memory safety by
providing the well-defined and predictable behavior by way of the execution
environment.

> ### _Memory safe language_

Reviewer 4: ```suggestion
### _Memory-safe language_
```

A programming language with sufficient defenses against memory safety bugs for
them to not be a significant source of security vulnerabilities. This requires
_preventing_ vulnerabilities or _ensuring_ correctness; mitigation is not
sufficient to provide an adequate level of memory safety.

> We identify several key requirements for a language to be memory safe:

Reviewer 4: ```suggestion
We identify several key requirements for a language to be memory-safe:
```

    -   [_Safety bugs_](#safety-bugs)
    -   [_Initial bug_](#initial-bug)
    -   [_Fail-stop_](#fail-stop)
-   [Vulnerability Terminology](#vulnerability-terminology)
    -   [_Vulnerability_ or _security vulnerability_](#vulnerability-or-security-vulnerability)
    -   [_Vulnerability defense_](#vulnerability-defense)
        -   [_Detecting_](#detecting)
        -   [_Mitigating_](#mitigating)
        -   [_Preventing_ vulnerabilities](#preventing-vulnerabilities)
        -   [_Ensuring_ correctness](#ensuring-correctness)
        -   [_Hardening_](#hardening)
-   [Memory Safety Specifics](#memory-safety-specifics)
    -   [_Memory safety_](#memory-safety)
        -   [_Temporal safety_](#temporal-safety)
        -   [_Spatial safety_](#spatial-safety)
        -   [_Type safety_](#type-safety)
        -   [_Initialization safety_](#initialization-safety)
        -   [_Data-race safety_](#data-race-safety)
    -   [_Memory safety bug_](#memory-safety-bug)
    -   [_Memory-safe platform_ or _environment_](#memory-safe-platform-or-environment)
    -   [_Memory-safe language_](#memory-safe-language)

<!-- tocstop -->

## Core Terminology

> ### _Hazard_

Reviewer 3: This term doesn't appear to be used anywhere. (The word hazard [appears once](https://github.com/carbon-language/carbon-lang/blob/629f77eb61b5f816f1e814ef68863d388fa8c870/proposals/p2868.md?plain=1#L108) in our repository, but not with this meaning.)

Author: The terminology set comes from a working draft I've been iterating on with some security teams, and they wanted to include this. It seems useful to stake out useful terminology here for future paper authors though?

Reviewer 3: I think it's fine to include this if we expect that it'll be used later, or if we want to use the term conversationally even though it doesn't appear in this proposal. But I went looking for an example of this term being used to try to better understand how to use it and didn't find one here :)

Could you maybe include an example in this section of a use of this term?

> For example, indexing an array with a user-supplied and unvalidated index is a hazard.

Author: Done.

Unsafe coding construct that may lead to a bug or vulnerability.

### _Bug_ or _defect_

Reachable program behavior contrary to the author's intent.

#### _Active bug_

Buggy behavior that is actively occurring for users of the program.

#### _Latent bug_

Buggy behavior that does not currently occur for users, but is reachable.
Behaviors that are reachable, and so _can_ happen, but don't happen today in
practice _are always still bugs_!

### _Safety_

> Absent a qualifier or narrow context, refers to _system safety_, and _safety

Reviewer 3: "system safety" and "safety engineering" don't seem to be defined here, making it unclear what we are defining "safety" to mean.

Author: Doh, my wikipedia links got lost somewhere. Fixed.

### _Safety bugs_

Bugs where some aspect of program behavior has insufficient (often none)
invariants or limits.

For example, _undefined behavior_ definitionally has no invariant or limit, and
reaching it is always a safety bug.

### _Initial bug_

The first behavior contrary to the author's intent, distinct from subsequent
deviations.

### _Fail-stop_

The behavior of immediately terminating the program, minimizing any further
business logic. This is in contrast to any form of "correct" program
termination, continuing execution, or unwinding.

## Vulnerability Terminology

### _Vulnerability_ or _security vulnerability_

A bug that creates the possibility for a malicious actor to subvert a program's
intended behavior in a way that violates a security policy (for example,
confidentiality, integrity, availability). Vulnerabilities are often exploitable
manifestations of underlying bugs.

### _Vulnerability defense_

The set of strategies and techniques employed to reduce the risks posed by
vulnerabilities arising from bugs. These strategies operate at different levels
and have varying degrees of effectiveness.

#### _Detecting_

> While still vulnerable, detecting or tracking the exploit of a bug. Requires

Reviewer 3: This seems to define "detecting" in terms of itself.

Author: Tried to improve.

I'm trying to define the term "detecting" in the context of vulnerability defense in terms of the English language definition of the word, but I see why that's confusing so rotated to different words and a bit more context.

_some_ invariant or limit, but very minimal.

#### _Mitigating_

Making a vulnerability significantly more expensive, difficult, or improbable to
be exploited.

#### _Preventing_ vulnerabilities

While still a bug, making it impossible to be a vulnerability. Often this is
done by defining behavior to [fail-stop](#fail-stop).

#### _Ensuring_ correctness

> No longer a bug, much less a vulnerability.

Reviewer 3: These two seem like incomplete definitions when read by themselves (which I expect these entries typically will be).

Author: Sure, added some detail.

#### _Hardening_

Combinations of mitigation, prevention, and ensured correctness to reduce the
practical risk of vulnerabilities due to bugs.

## Memory Safety Specifics

### _Memory safety_

Having well-defined and predictable behavior regarding memory access, even in

> the face of bugs. Memory safety encompass several key aspects:

Reviewer 3: ```suggestion
the face of bugs. Memory safety encompasses several key aspects:
```
